### DISCUSSION TITLE
Anti-Opmode

### DISCUSSION THREAD

**Discussion by guguxh** [wxyz-abcd/node-haxball#51] [General]
How to check the frame send to the host? And if this is more than the boundlimit, check too

**Comment by wxyz-abcd** [comment]
The callback `onOperationReceived` has been updated with the signature: `onOperationReceived(obj, msg, globalFrameNo, clientFrameNo, customData)` where `globalFrameNo` is the host's frame number value, and `clientFrameNo` is the frame number that was received with the message from the current client and is being sent to all clients with the message. The method that we used is as follows:

  - Filter the messages to only check the input event messages. 
  - Subtract globalFrameNo from clientFrameNo and compare the result with several values to detect cheat users. (3, 4, 5 seem to be acceptable values, sometimes 2 is also acceptable.) 
  - If the difference is 2, check whether there are more than 7 2's in the latest 20 difference values calculated. (Of course, this is done seperately for each user.)
  
However, the tests were inconclusive and we got many false-positives. The problem is that the difference value seems to be affected by the performance of client. (or maybe numerous other things) Currently I'm finishing web examples, but I will get back to it sometime and hopefully we will find a better solution.
